src/ui/service/token/creator.go

```diff
@@ -0,0 +1,237 @@
+/*
+   Copyright (c) 2016 VMware, Inc. All Rights Reserved.
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package token
+
+import (
+	"fmt"
+	"github.com/docker/distribution/registry/auth/token"
+	"github.com/vmware/harbor/src/common/dao"
+	"github.com/vmware/harbor/src/common/utils/log"
+	"github.com/vmware/harbor/src/ui/config"
+	"net/http"
+	"strings"
+)
+
+var creatorMap map[string]Creator
+
+const (
+	notary   = "harbor-notary"
+	registry = "harbor-registry"
+)
+
+//InitCreators initialize the token creators for different services
+func InitCreators() {
+	creatorMap = make(map[string]Creator)
+	ext, err := config.ExtEndpoint()
+	if err != nil {
+		log.Warningf("Failed to get ext enpoint, err: %v, the token service will not be functional with notary requests", err)
+	} else {
+		creatorMap[notary] = &generalCreator{
+			validators: []ReqValidator{
+				&basicAuthValidator{},
+			},
+			service: notary,
+			filterMap: map[string]accessFilter{
+				"repository": &repositoryFilter{
+					parser: &endpointParser{
+						endpoint: strings.Split(ext, "//")[1],
+					},
+				},
+			},
+		}
+	}
+
+	creatorMap[registry] = &generalCreator{
+		validators: []ReqValidator{
+			&secretValidator{config.JobserviceSecret()},
+			&basicAuthValidator{},
+		},
+		service: registry,
+		filterMap: map[string]accessFilter{
+			"repository": &repositoryFilter{
+				//Workaround, had to use same service for both notary and registry
+				parser: &endpointParser{
+					endpoint: ext,
+				},
+			},
+			"registry": &registryFilter{},
+		},
+	}
+}
+
+// Creator creates a token ready to be served based on the http request.
+type Creator interface {
+	Create(r *http.Request) (*tokenJSON, error)
+}
+
+type imageParser interface {
+	parse(s string) (*image, error)
+}
+
+type image struct {
+	namespace string
+	repo      string
+	tag       string
+}
+
+type basicParser struct{}
+
+func (b basicParser) parse(s string) (*image, error) {
+	return parseImg(s)
+}
+
+type endpointParser struct {
+	endpoint string
+}
+
+func (e endpointParser) parse(s string) (*image, error) {
+	repo := strings.SplitN(s, "/", 2)
+	if len(repo) < 2 {
+		return nil, fmt.Errorf("Unable to parse image from string: %s", s)
+	}
+	//Workaround, need to use endpoint Parser to handle both cases.
+	if strings.ContainsRune(repo[0], '.') {
```

Comment:
This is the first chunk, the parser makes best effort to guess if it starts with a host name.  I think docker has the same logic?